File: language/src/main/java/brj/nodes/FnNode.java
package brj.nodes;

import brj.BridjeLanguage;
import brj.runtime.BridjeFunction;
import com.oracle.truffle.api.dsl.NodeField;
import com.oracle.truffle.api.dsl.Specialization;
import com.oracle.truffle.api.source.SourceSection;
import org.jetbrains.annotations.NotNull;

@NodeField(name = "fn", type = BridjeFunction.class)
@NodeField(name = "sourceSection", type = SourceSection.class)
public abstract class FnNode extends ExprNode {
    public FnNode(BridjeLanguage lang) {
        super(lang);
    }

    protected abstract BridjeFunction getFn();

    @NotNull
    @Specialization
    public Object doExecute() {
        return getFn();
    }
}

File: language/src/main/kotlin/brj/nodes/FnNode.kt
package brj.nodes

import brj.BridjeNode
import brj.runtime.BridjeFunction
import brj.runtime.ClosureBridjeFunction
import com.oracle.truffle.api.RootCallTarget
import com.oracle.truffle.api.frame.VirtualFrame
import com.oracle.truffle.api.source.SourceSection

class FnNode(
    private val function: BridjeFunction,
    loc: SourceSection? = null
) : BridjeNode(loc) {
    override fun execute(frame: VirtualFrame): Any = function
}

class ClosureFnNode(
    private val callTarget: RootCallTarget,
    private val captureSlots: IntArray,
    loc: SourceSection? = null
) : BridjeNode(loc) {
    override fun execute(frame: VirtualFrame): Any {
        val capturedValues = Array<Any?>(captureSlots.size) { i ->
            frame.getObject(captureSlots[i])
        }
        return ClosureBridjeFunction(callTarget, capturedValues)
    }
}
